src/dbnode/persist/fs/merger.go

```diff
@@ -0,0 +1,303 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fs
+
+import (
+	"io"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/digest"
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/persist"
+	"github.com/m3db/m3/src/dbnode/storage/index/convert"
+	"github.com/m3db/m3/src/dbnode/ts"
+	"github.com/m3db/m3/src/dbnode/x/xio"
+	"github.com/m3db/m3/src/x/checked"
+	"github.com/m3db/m3/src/x/context"
+	"github.com/m3db/m3/src/x/ident"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+type merger struct {
+	reader         DataFileSetReader
+	blockAllocSize int
+	srPool         xio.SegmentReaderPool
+	multiIterPool  encoding.MultiReaderIteratorPool
+	identPool      ident.Pool
+	encoderPool    encoding.EncoderPool
+	nsOpts         namespace.Options
+}
+
+// NewMerger returns a new Merger. This implementation is in charge of merging
+// the data from an existing fileset with a merge target. If data for a series
+// at a timestamp exists both on disk and the merge target, data from the merge
+// target will be used. This merged data is then persisted.
+//
+// Note that the merger does not know how or where this merged data is
+// persisted since it just uses the flushPreparer that is passed in. Further,
+// it does not signal to the database of the existence of the newly persisted
+// data, nor does it clean up the original fileset.
+func NewMerger(
+	reader DataFileSetReader,
+	blockAllocSize int,
+	srPool xio.SegmentReaderPool,
+	multiIterPool encoding.MultiReaderIteratorPool,
+	identPool ident.Pool,
+	encoderPool encoding.EncoderPool,
+	nsOpts namespace.Options,
+) Merger {
+	return &merger{
+		reader:         reader,
+		blockAllocSize: blockAllocSize,
+		srPool:         srPool,
+		multiIterPool:  multiIterPool,
+		identPool:      identPool,
+		encoderPool:    encoderPool,
+		nsOpts:         nsOpts,
+	}
+}
+
+// Merge merges data from a fileset with a merge target and persists it.
+// The caller is responsible for finalizing all resources used for the
+// MergeWith passed here.
```

Comment:
Can you document somewhere (here or on `ForEachRemaining`) that while the data passed to `ForEachRemaining` is basically a copy that will be finalized when the context is closed, the ID and tags are expected to live for as long as the user of the MergeWith requires them so they should either be `NoFinalize()` or passed as copies (in your case since you're taking them from the shard/series they're NoFinalize() and its safe but would like to document)